python/pyspark/sql/tests/test_functions.py

```diff
@@ -1268,6 +1268,12 @@ def test_bucket(self):
             message_parameters={"arg_name": "numBuckets", "arg_type": "str"},
         )
 
+    def test_no_cast(self):
```

Comment:
How about add a `test_cast`?

Seems like the test for `functions.cast` is missing (And that's also the root reason why we haven't noticed this is wrong until now)

Reply:
I apologize, but I am not sure I understand what you mean... it sounds like you think that `functions.cast` already exists.
It does not, there is only `Column.cast`.

We _could_ add a `functions.cast` method, but I don't think we should because 
1. the need is already covered by `Column.cast`
2. that would require to add it to the Scala/Java API to be consistent

Reply:
Oh,, yeah I thought we have `functions.cast`.

Hmm... If `functions.cast` does not exist, I'm not sure if removing the duplication of `typing.cast` in advance would be appropriate.

WDYT, @HyukjinKwon ?

Reply:
From what I could test, the current behavior when using `typing.cast` is that it has no effect. The SQL cast does not happen, obviously, but no exception is raised either.
Removing `typing.cast` will break jobs that are currently using it. This can be seen as a regression, but since it breaks a buggy behaviour, it could also be seen as a bugfix ?

I see 4 possible alternatives here:
1. Remove `typing.cast` (the current commit)
2. Add a Python version of `functions.cast` with the right behaviour that uses `Column.cast`
3. Replace the `typing.cast` call with a dummy `functions.cast` that logs a warning
4. Add `functions.cast` to the Scala/Java API then the Python one (and probably R too ?)

All options have their upsides and downsides:

1. Already implemented, consistent with Scala API, but will crash jobs that mistakenly used `functions.cast`
2. Simple to implement, inconsistent with Scala API, might silently change the behaviour of jobs that mistakenly used `functions.cast` (because the behaviour will change from "no effect" to "actual cast")
3. Simple to implement, consistent with Scala API, will not crash jobs, only issue warnings (that users might or might not see)
4. Longest to implement, consistent with Scala API, might silently change the behaviour of jobs that mistakenly used `functions.cast` (because the behaviour will change from "no effect" to "actual cast")

Personally, I think 1. is okay, but it's your choice.

Reply:
I don't think there's a concern for the name collision with `typing.cast` because `functions.cast` is not a function specified in the [PySpark API Reference](https://spark.apache.org/docs/latest/api/python/reference/pyspark.sql/functions.html).

However, I also sort of agree that users might get confused and expect a `functions.cast` to exist since it's a function name that is generally expected to be supported (and anyway it is also supported in the Column API...) based on the example you provided in the PR description.

At the moment, I am not 100% sure which direction is the right based on my own thoughts, so let's gather some more opinions from others.

WDYT, @HyukjinKwon @ueshin @zhenglaizhang ??